src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.Editor.cs

```diff
@@ -0,0 +1,255 @@
+    {
+        /// <summary>
+        /// We perform 3 different kinds of edits for Move Type Code Action
+        /// 1. Rename file to match type
+        /// 2. Rename type to match file
+        /// 3. Move type to a new file
+        /// </summary>
+        private partial class Editor
+        {
+            private readonly CancellationToken _cancellationToken;
+            private readonly State _state;
+            private readonly TService _service;
+
+            private readonly bool _renameFile;
+            private readonly bool _renameType;
+
+            public Editor(
+                TService service,
+                State state,
+                bool renameFile,
+                bool renameType,
+                CancellationToken cancellationToken)
+            {
+                _renameFile = renameFile;
+                _renameType = renameType;
+                _state = state;
+                _service = service;
+                _cancellationToken = cancellationToken;
+            }
+
+            private SemanticDocument SemanticDocument => _state.SemanticDocument;
+
+            /// <summary>
+            /// operations performed by Move Type CodeAction.
+            /// </summary>
+            internal async Task<IEnumerable<CodeActionOperation>> GetOperationsAsync()
+            {
+                var solution = SemanticDocument.Document.Project.Solution;
+
+                if (_renameFile)
+                {
+                    return RenameFileToMatchTypeName(solution);
+                }
+                else if (_renameType)
+                {
+                    return await RenameTypeToMatchFileAsync(solution).ConfigureAwait(false);
+                }
+
+                return await MoveTypeToNewFileAsync(_state.TargetFileNameCandidate).ConfigureAwait(false);
+            }
+
+            /// <summary>
+            /// Given a document and a type contained in it, moves the type
+            /// out to its own document. The new document's name typically
+            /// is the type name, or is atleast based on the type name.
+            /// </summary>
+            /// <param name="documentName">name for the new document being added</param>
+            /// <remarks>
+            /// The algorithm for this, is as follows:
+            /// 1. Fork the original document that contains the type to be moved.
+            /// 2. Keep the type and required namespace containers, using statements
+            ///     and remove everything else from the forked document.
+            /// 3. Add this forked document to the solution.
+            /// 4. Finally, update the original document and remove the type from it.
+            /// </remarks>
+            private async Task<IEnumerable<CodeActionOperation>> MoveTypeToNewFileAsync(string documentName)
+            {
+                // Fork, update and add as new document.
+                var projectToBeUpdated = SemanticDocument.Document.Project;
```

Comment:
assert that the new document name isn't the same as the existing document you're forking.